rfd/0047-repository-management-features.md

```diff
@@ -0,0 +1,274 @@
+
+**Status:** _In Progress_ 
+
+###  What 
+
+Automatically backport changes to `docs/` to the specified branches. 
+
+### Why 
+
+Backporting docs is time consuming and there are often many doc changes. Implementing a way to automate backporting will increase the efficiency and speed of the development team. 
+
+### Implementation
+
+Command: `backport`
+
+
+#### Informing the Bot
+
+When there is a pull request opened and needs to be backported, the author will tag the `github-actions` bot with the string "Backport" and following will be a comma separated list of the branches to backport to.  Example backport comment: 
+
+``` 
+ @github-actions Backport branch-1, branch-2
+```
+
+#### Backport Requests
+
+The workflow will trigger anytime the `issue_comment` event occurs. 
+
+The `issue_comment` event occurs for comments on both issues and pull requests. To determine whether the `issue_comment` event was triggered from a pull request, the event payload will have an  `issue.pull_request` property and we can use it as a condition to only run the workflow on pull requests comment events.
+
+```yaml
+# docs-backport.yml
+on:
+  issue_comment:
+    types: [created]
+  jobs:
+    merge_job:
+      # This will only run for pull request comments
+      if: ${{ github.event.issue.pull_request }}
+      runs-on: ubuntu-latest
+      # jobs...
+            
+```
+The backport bot will only support merged pull requests. 
+
+
+#### Backporting 
+
+The process of backporting a branch is checking out the targeted branch and using `git cherry-pick` with the merge commit. The following process will happen for every branch label in the pull request. The bot will use the `os/exec` package to checkout branches and `cherry-pick` the commit.  
+
+_A potential library to use: [`git2go`](https://github.com/libgit2/git2go) which is essentially a Go wrapper for `libgit2`._
+
+
+```go
+    // Example use of `os/exec`
+    // Will be used with `cherry-pick` command, too. 
+    cmd := exec.Command("git", "checkout", branchName)
+    if err := cmd.Run(); err != nil {
+        if werr, ok := err.(*exec.ExitError); ok {
+            if s := werr.Error(); s != "0" {
+                // non-zero exit code
+                // merge is not clean
+            }
+        }
+    }
+```
+
+If the merge is clean, the changes will be merged directly into the branch. If not, a comment will be posted on the pull request that requested the backport stating that the bot was unable to backport cleanly. The bot will fail and the author will need to manually backport their changes. 
+
+
+Example comment: 
+```
+Failed to backport changes to <branch>, <branch>, <branch>.
+```
+
+#### Permissions 
+
+To support automatic backporting, `contents:write` and `issues:write` permissions will need to be granted to the Github token. Write access to `contents` is needed for merging the changes and write access to issues will allow posting comments on pull requests in the event of a failed backport.
```

Comment:
This does open us up to attack quite a bit right?

Reply:
I wonder if an alternative would be to try and create a PR with the changes, rather than writing them directly to the target branch. Not sure how much value that leaves for the backport bot, though.

Reply:
@russjones Yeah, if an attacker were to get the Github token, they could merge anything in. 

@tcsc I considered just opening up a pull request and yeah there wouldn't be much value. The time saved would just be backporting, but then a contributor would have to babysit a PR and merge it in themselves.

Reply:
I don't think we should give the bot permissions to create commits in our repo. If we do that that would let an attacker write to our repo.

Instead as a first step let's automatically create a backport PR. We can come back and look at this later.